### PR TITLE
Leap 15 enablement jsc#PED-3258

### PR DESCRIPTION
**Pull request by lkocman**
- Backport SLE Micro with_selinux changes 
  - extra handling of cockpit-ws / selinux related docs
- add 0004-leap-gnu11.patch as gnu18 is not available on Leap

### CODE REVIEW

**Comment by lkocman** [comment]
Submitting changes from https://build.opensuse.org/package/show/SUSE:SLE-15-SP4:Update:Products:Micro54/cockpit + enablement on Leap 15.4+ (where we have no SELinux policy, just like in SLES).

I quite like the with_selinux, I think that's worth considering even for upstream.

I did discuss the policy enablement on Leap with @jsegitz but he mentioned that team has no capacity for it.

Leap 15.4 doesn't get build as there seems to be a problem with tar_scm service (the obscpio for openSUSE theme is somehow missed). We plan to add it to 15.6 only, unless PM would vote for enablement in SLES (see https://jira.suse.com/browse/PED-3258).

**Comment by lkocman** [comment]
Attaching zipped patch (github doesn't allow me to attach the .patch file directly)
[0004-leap-gnu11.zip](https://github.com/openSUSE/cockpit/files/11735656/0004-leap-gnu11.zip)

I suppose that patches are kept in OBS, as I coudln't find them within checkout.

File: tools/cockpit.spec [hunk 1]
```diff
@@ -219,6 +233,10 @@ BuildRequires:  python3-tox-current-env
 %if 0%{?is_smo}
 %patch101 -p1
 %patch102 -p1
+%else
```
Comment (AdamMajer) [inline tools/cockpit.spec:236]: no need %else here, I think

File: tools/cockpit.spec [hunk 2]
```diff
@@ -69,6 +69,15 @@ Patch5:         storage-btrfs.patch
 Patch101:       hide-pcp.patch
 Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
 
+# Leap specific patch
+
+%if 0%{?is_smo}
+# Ensure we do not affect Micro with SLE/Leap specifi patches
+%else
+%if 0%{?sle_version} >= 150400 && 0%{?sle_version} <= 150700
```
Comment (AdamMajer) [inline tools/cockpit.spec:77]: can remove the %if here, just add patches... maybe start with number 200 for Leap, so they don't get mixed up

File: tools/cockpit.spec [hunk 3]
```diff
@@ -235,9 +253,10 @@ echo "m4_define(VERSION_NUMBER, %version)" > version.m4
 autoreconf -fvi -I tools
 #
 %configure \
-    %{?selinux_configure_arg} \
+    --disable-silent-rules \
```
Comment (AdamMajer) [inline tools/cockpit.spec:256]: these 2 lines should remain as-is. the selinux one we need for selinux, and disable-silent-rules could be different patch that could be upstreamed.

File: tools/cockpit.spec [hunk 4]
```diff
@@ -235,9 +253,10 @@ echo "m4_define(VERSION_NUMBER, %version)" > version.m4
...
     --with-cockpit-user=cockpit-ws \
     --with-cockpit-ws-instance-user=cockpit-wsinstance \
+    --with-selinux-config-type=etc_t \
```
Comment (AdamMajer) [inline tools/cockpit.spec:259]: different patch from this one

File: tools/cockpit.spec [hunk 5]
```diff
@@ -249,10 +268,12 @@ autoreconf -fvi -I tools
     --disable-ssh \
 %endif
 
-make -f /usr/share/selinux/devel/Makefile cockpit.pp
-bzip2 -9 cockpit.pp
+make -j4 %{?extra_flags} all
```
Comment (AdamMajer) [inline tools/cockpit.spec:271]: wrap the above without reformatting so least changes.

%make_build should work with leap.

File: tools/cockpit.spec [hunk 6]
```diff
@@ -275,12 +296,13 @@ install -p -m 644 tools/cockpit.pam $RPM_BUILD_ROOT%{_sysconfdir}/pam.d/cockpit
...
-install -D -m 644 -t %{buildroot}%{_mandir}/man8 selinux/%{name}_ws_selinux.8cockpit
-# create this directory in the build root so that %ghost sees the desired mode
-install -d -m 700 %{buildroot}%{_sharedstatedir}/selinux/%{selinuxtype}/active/modules/200/%{name}
+%if 0%{?with_selinux}
+    install -D -m 644 %{name}.pp.bz2 %{buildroot}%{_datadir}/selinux/packages/%{selinuxtype}/%{name}.pp.bz2
+    install -D -m 644 -t %{buildroot}%{_mandir}/man8 selinux/%{name}_session_selinux.8cockpit
+    install -D -m 644 -t %{buildroot}%{_mandir}/man8 selinux/%{name}_ws_selinux.8cockpit
+    # create this directory in the build root so that %ghost sees the desired mode
+    install -d -m 700 %{buildroot}%{_sharedstatedir}/selinux/%{selinuxtype}/active/modules/200/%{name}
+%endif
```
Comment (AdamMajer) [inline tools/cockpit.spec:305]: same here, no reformatting for lines so this can be re-applied later

**Review by AdamMajer — CHANGES_REQUESTED** [review]

File: tools/cockpit.spec (hunk 1 above)
Comment (lkocman) [reply tools/cockpit.spec:236]: Here I think we want to check just for SLES 15 SPX codebase. That will cover SMO too.

Comment (AdamMajer) [reply tools/cockpit.spec:236]: I mean insstead of

```
%else
%if 0%{?sle_version} >= 150400 && 0%{?sle_version} <= 150700
%patch103 -p1
%endif
%endif

```

just have

```
%if 0%{?sle_version} >= 150400 && 0%{?sle_version} <= 150700
%patch103 -p1
%endif
```

Or even to simplify this to

`%if 0%{?suse_version} == 1500` because this problem also affects smo, they just don't know it yet ;)

File: tools/cockpit.spec (hunk 2 above)
Comment (lkocman) [reply tools/cockpit.spec:77]: we want this for both in fact

File: tools/cockpit.spec (hunk 4 above)
Comment (lkocman) [reply tools/cockpit.spec:259]: removed

File: tools/cockpit.spec (hunk 5 above)
Comment (lkocman) [reply tools/cockpit.spec:271]: restored

File: tools/cockpit.spec (hunk 6 above)
Comment (lkocman) [reply tools/cockpit.spec:305]: removed formatting

File: tools/cockpit.spec (hunk 5 above)
Comment (lkocman) [reply tools/cockpit.spec:271]: fixed

File: tools/cockpit.spec (hunk 3 above)
Comment (lkocman) [reply tools/cockpit.spec]: fixed

File: tools/cockpit.spec [hunk 7]
```diff
@@ -107,10 +111,13 @@ Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
...
 # Ship custom SELinux policy (but not for cockpit-appstream)
-%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version}
+# SLES / Leap 15 does not have selinux-policy. TW and SLE Micro does.
+%define with_selinux 0
+%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version} > 1600 || 0%{?is_smo}
 %if "%{name}" == "cockpit"
 %define selinuxtype targeted
-%define selinux_configure_arg --enable-selinux-policy=%{selinuxtype}
+%define with_selinux 1
+%define selinux_policy_version %(rpm --quiet -q selinux-policy && rpm -q --queryformat "%{V}-%{R}" selinux-policy || echo 1)
```
Comment (AdamMajer) [inline tools/cockpit.spec:120]: the `%_selinux_policy_version` comes from `selinux-policy`, so there is no need to redefine it here or to remove the underscore, especially since it only applies to case with selinux

best to remove this line

File: tools/cockpit.spec (hunk 2 above)
Comment (AdamMajer) [reply tools/cockpit.spec:73]: OK, but remove the %if wrapper completely. We have the %if wrapper in the %setup around %patch

File: tools/cockpit.spec [hunk 8]
```diff
@@ -540,8 +557,10 @@ Summary: Cockpit Web Service
 Requires: glib-networking
 Requires: openssl
 Requires: glib2 >= 2.50.0
-Requires: (selinux-policy >= %{_selinux_policy_version} if selinux-policy-%{selinuxtype})
+%if 0%{?with_selinux}
+Requires: (selinux-policy >= %{selinux_policy_version} if selinux-policy-%{selinuxtype})
```
Comment (AdamMajer) [inline tools/cockpit.spec:561]: re-add the _ to `%{selinux_policy_version}`

File: tools/cockpit.spec [hunk 9]
```diff
@@ -224,6 +233,10 @@ BuildRequires:  python3-tox-current-env
 %patch101 -p1
 %patch102 -p1
 %endif
+# For anything based on SLES 15 codebase (including Leap, SLEM)
+%if 0%{?sle_version} >= 150400 && 0%{?sle_version} <= 150700
```
Comment (AdamMajer) [inline tools/cockpit.spec:237]: If we just have `%if 0%{?suse_version} == 1500`, then it should be sufficient. We don't care about out of support (or soon to be) Leaps and SLES and if they don't build.

File: tools/cockpit.spec [hunk 10]
```diff
@@ -107,10 +111,13 @@ Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
 %endif
 
 # Ship custom SELinux policy (but not for cockpit-appstream)
-%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version}
+# SLES / Leap 15 does not have selinux-policy. TW and SLE Micro does.
+%define with_selinux 0
+%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version} > 1600 || 0%{?is_smo}
 %if "%{name}" == "cockpit"
 %define selinuxtype targeted
-%define selinux_configure_arg --enable-selinux-policy=%{selinuxtype}
```
Comment (AdamMajer) [inline tools/cockpit.spec:113]: line should be be re-added

File: tools/cockpit.spec [hunk 11]
```diff
@@ -107,10 +111,13 @@ Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
 %endif
 
 # Ship custom SELinux policy (but not for cockpit-appstream)
-%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version}
+# SLES / Leap 15 does not have selinux-policy. TW and SLE Micro does.
+%define with_selinux 0
```
Comment (AdamMajer) [inline tools/cockpit.spec:115]: These 2 lines are not needed. Everywhere the guards are

`%if 0%{?with_selinux}`

which means by default it will build without selinux. Defining it to 0 is not necessary here.

**Review by AdamMajer — CHANGES_REQUESTED** [review]
Patch is getting smaller :-) I want to make it smaller still!

File: tools/cockpit.spec (hunk 7 above)
Comment (lkocman) [reply tools/cockpit.spec:120]: removed

File: tools/cockpit.spec (hunk 8 above)
Comment (lkocman) [reply tools/cockpit.spec:561]: fixed

File: tools/cockpit.spec (hunk 10 above)
Comment (lkocman) [reply tools/cockpit.spec:113]: fixed

File: tools/cockpit.spec (hunk 11 above)
Comment (lkocman) [reply tools/cockpit.spec]: removed

File: tools/cockpit.spec (hunk 9 above)
Comment (lkocman) [reply tools/cockpit.spec]: Yes, that makes sense. I wanted to avoid < than 15 SP4, as in older streams we still don't have anything newer than python3.6 (aside from minimal 3.10 etc)  I believe that was also one of blockers for newer cockpits.

File: tools/cockpit.spec [hunk 12]
```diff
@@ -107,10 +111,11 @@ Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
 %endif
 
 # Ship custom SELinux policy (but not for cockpit-appstream)
-%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version}
+%if 0%{?rhel} >= 9 || 0%{?fedora} || 0%{?suse_version} > 1600 || 0%{?is_smo}
```
Comment (AdamMajer) [inline tools/cockpit.spec:114]: This needs to be >= 1600 since ALP is 1600 :wink:

File: tools/cockpit.spec [hunk 13]
```diff
@@ -655,7 +674,6 @@ if [ "$1" = 2 ]; then
     test -f $certfile && stat -c '%G' $certfile | grep -q cockpit-wsinstance && chgrp cockpit-ws $certfile
 fi
 
-%if 0%{?suse_version}
```
Comment (AdamMajer) [inline tools/cockpit.spec]: should be re-added

File: tools/cockpit.spec [hunk 14]
```diff
@@ -69,6 +69,10 @@ Patch5:         storage-btrfs.patch
 Patch101:       hide-pcp.patch
 Patch102:       0002-selinux-temporary-remove-setroubleshoot-section.patch
 
+# For anything based on SLES 15 codebase (including Leap, SLE Micro)
+%if 0%{?suse_version} == 1500
```
Comment (AdamMajer) [inline tools/cockpit.spec:73]: You don't need to wrap Patch: with %if. If the patches are unapplied, it's ok. But we can't have them missing because then we have errors with source validators at `osc commit` times.

**Review by AdamMajer — CHANGES_REQUESTED** [review]

**Review by AdamMajer — APPROVED** [review]

**Comment by lkocman** [comment]
https://build.opensuse.org/request/show/1093337Thank you! Merging
